pkg/kv/kvserver/rangefeed/scheduler.go

```diff
@@ -0,0 +1,494 @@
+			// lock ourselves.
+			close(s.drained)
+		} else {
+			for id := range s.mu.fs {
+				_, _ = s.enqueueInternalLocked(id, stopped)
+			}
+		}
+	}
+	s.mu.Unlock()
+	s.cond.Broadcast()
+
+	// Now wait for all processors to drain and unregister.
+	func() {
+		for {
+			select {
+			case <-s.drained:
+				return
+			case <-time.After(10 * time.Second):
+				s.mu.Lock()
+				registered := len(s.mu.fs)
+				s.mu.Unlock()
+				log.Infof(context.Background(), "schedule shutdown: active processors count: %d", registered)
+			}
+		}
+	}()
+
+	// Stop all worker threads.
+	s.mu.Lock()
+	s.mu.stopped = true
+	s.mu.Unlock()
+	s.cond.Broadcast()
+
+	s.wg.Wait()
+}
+
+// ClientScheduler is a wrapper on top of scheduler that could be passed to a
+// processor to be able to register itself with a pre-configured ID, enqueue
+// events and terminate as needed.
+type ClientScheduler struct {
+	id int64
+	s  *Scheduler
+}
+
+// NewClientScheduler creates an instance of ClientScheduler for specific id.
+// It is safe to use it as value as it is immutable and delegates all work to
+// underlying scheduler.
+func NewClientScheduler(id int64, s *Scheduler) ClientScheduler {
+	return ClientScheduler{
+		id: id,
+		s:  s,
+	}
+}
+
+// ID returns underlying callback id used to schedule work.
+func (cs *ClientScheduler) ID() int64 {
+	return cs.id
+}
+
+// Register registers processing callback in scheduler. Error is returned if
+// callback was already registered for this ClientScheduler or if scheduler is
+// already quiescing.
+func (cs *ClientScheduler) Register(cb Callback) error {
+	return cs.s.Register(cs.id, cb)
+}
+
+// Schedule schedules callback for event. Error is returned if client callback
+// wasn't registered prior to this call.
+func (cs *ClientScheduler) Schedule(event processorEventType) error {
+	return cs.s.Enqueue(cs.id, event)
+}
+
+// Stop instructs processor to stop gracefully by sending it stopped event.
+// Once stop is called all subsequent Schedule calls will return error.
+func (cs *ClientScheduler) Stop() {
+	cs.s.Stop(cs.id)
+}
+
+// Unregister will remove callback associated with this processor. No stopped
+// event will be scheduled. See Scheduler.Unregister for details.
+func (cs *ClientScheduler) Unregister() {
+	cs.s.Unregister(cs.id)
+}
+
+// Number of queue elements allocated at once to amortize queue allocations.
+const idQueueChunkSize = 1000
+
+type idQueueChunk struct {
+	data [idQueueChunkSize]int64
+	next *idQueueChunk
+}
+
+// idQueue stores pending processor ID's. Internally data is stored in
+// idQueueChunkSize sized arrays that are added as needed and discarded once
+// reader and writers finish working with it. Since we only have a single
+// scheduler per store, we don't use a pool as only reuse could happen within
+// the same queue and in that case we can just increase chunk size.
+type idQueue struct {
+	first, last *idQueueChunk
+	read, write int
+	size        int
+}
+
+func newIDQueue() *idQueue {
```

Comment:
would be nice to move this to separate file w/ dedicated tests (no matter how trivial it is).